cli/download_source_test.go

```diff
@@ -131,17 +131,32 @@ func TestDownloadTerraformSourceIfNecessaryRemoteUrlToAlreadyDownloadedDirSameVe
 
 	copyFolder(t, "../test/fixture-download-source/hello-world-version-remote", downloadDir)
 
-	testDownloadTerraformSourceIfNecessary(t, canonicalUrl, downloadDir, "# Hello, World version remote")
+	testDownloadTerraformSourceIfNecessary(t, canonicalUrl, downloadDir, false, "# Hello, World version remote")
 }
 
-func testDownloadTerraformSourceIfNecessary(t *testing.T, canonicalUrl string, downloadDir string, expectedFileContents string) {
+func TestDownloadTerraformSourceIfNecessaryRemoteUrlOverrideSource(t *testing.T) {
+	t.Parallel()
+
+	canonicalUrl := "github.com/gruntwork-io/terragrunt//test/fixture-download-source/hello-world?ref=v0.9.7"
+	downloadDir := tmpDir(t)
+	defer os.Remove(downloadDir)
+
+	copyFolder(t, "../test/fixture-download-source/hello-world-version-remote", downloadDir)
+
+	testDownloadTerraformSourceIfNecessary(t, canonicalUrl, downloadDir, true, "# Hello, World")
+}
+
+func testDownloadTerraformSourceIfNecessary(t *testing.T, canonicalUrl string, downloadDir string, updateSource bool, expectedFileContents string) {
 	terraformSource := &TerraformSource{
 		CanonicalSourceURL: parseUrl(t, canonicalUrl),
 		DownloadDir: downloadDir,
 		VersionFile: filepath.Join(downloadDir, "version-file.txt"),
 	}
 
-	err := downloadTerraformSourceIfNecessary(terraformSource, options.NewTerragruntOptionsForTest("./should-not-be-used"))
+	terragruntOptions := options.NewTerragruntOptionsForTest("./should-not-be-used")
+	terragruntOptions.UpdateSource = updateSource
+
+	err := downloadTerraformSourceIfNecessary(terraformSource, terragruntOptions)
```

Comment:
This is a helper function that is used by a bunch of actual test cases up above. The `TestDownloadTerraformSourceIfNecessaryRemoteUrlToAlreadyDownloadedDirSameVersion` and `TestDownloadTerraformSourceIfNecessaryRemoteUrlOverrideSource` test cases both set the `source` URL to the same value and both run against folders that contain downloaded code of exactly the same version hash as the `source` URL. The difference is that the first test case doesn't set `UpdateSource`, so it verifies that nothing is downloaded, while the second test case does set `UpdateSource`, so it verifies that the new code is downloaded from the `source` URL.